polynote-server/src/main/scala/polynote/server/NotebookSession.scala

```diff
@@ -117,8 +117,37 @@ class NotebookSession(subscriber: KernelSubscriber, streamingHandles: StreamingH
         ZIO.foreach_(presence.flatMap(_._2.toList).map(sel => KernelStatus(sel)))(PublishMessage)
   }
 
+  // Send the results, along with some progress updates based on how many results there are
+  private def sendCellResults(results: List[CellResult], taskInfo: TaskInfo): ZIO[PublishMessage, Throwable, Unit] = {
+    val sendResults = ZIO.when(results.nonEmpty) {
+      (results.size / 10) match {
+        case 0 => ZIO.foreach_(results)(PublishMessage)
+        case chunkSize =>
+          val chunks = results.sliding(chunkSize, chunkSize).toList
+          val messages = chunks.head ::: chunks.tail.zipWithIndex.flatMap {
+            case (chunk, index) => KernelStatus(UpdatedTasks(List(taskInfo.progress(0.5 + 0.05 * index)))) :: chunk
+          }
+
+          ZIO.foreach_(messages)(PublishMessage)
+      }
+    }
+    sendResults *> PublishMessage(KernelStatus(UpdatedTasks(List(taskInfo.completed))))
+  }
+
+  // First send the notebook without any results (because they're large) and then send the individual results
+  // to make notebook loading more incremental.
+  private def sendNotebook: RIO[BaseEnv with GlobalEnv with PublishMessage, Unit] = for {
+    nb      <- subscriber.notebook
+    taskInfo = TaskInfo(nb.path, s"Loading notebook", s"Loading ${nb.path}", Running)
+    _       <- PublishMessage(KernelStatus(UpdatedTasks(List(taskInfo))))
+    _       <- PublishMessage(nb.withoutResults)
+    _       <- PublishMessage(KernelStatus(UpdatedTasks(List(taskInfo.progress(0.5)))))
+    results  = nb.cells.flatMap(cell => cell.results.map(CellResult(cell.id, _)))
```

Comment:
Not sure how often we do this but it seems like `nb.results` would be a good complement to `nb.withoutResults` (doesn't necessarily need to be in this PR).

Reply:
Good idea, did that